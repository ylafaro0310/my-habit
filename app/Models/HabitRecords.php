<?php

namespace App\Models;

use Illuminate\Support\Facades\DB;
use Illuminate\Support\Facades\Log;

class HabitRecords {
    public $validationRules = [
        'habit_id' => 'required|integer',
        'completed_at' => 'required|integer',
        'is_skipped' => 'required|boolean',
    ];
    
    function __construct(){
        $this->pdo = DB::connection()->getPdo();
        $this->table_name = 'habit_records';
    }

    public function index(){
        $sth = $this->pdo->prepare("select * from $this->table_name");
        $sth->execute();
        $result = $sth->fetchAll();
        Log::debug($result);
        return $result;
    }

    public function store($params){
        $columns = implode(', ',array_keys($params));
        $values = implode(', ',array_map(function($v){return "'$v'";},array_values($params)));
        $sth = $this->pdo->prepare("insert into $this->table_name (${columns}) values (${values})");
        return $sth->execute();
    }

    public function destroy($id){
        $sth = $this->pdo->prepare("delete from $this->table_name where id = ${id}");
        return $sth->execute();
    }
}