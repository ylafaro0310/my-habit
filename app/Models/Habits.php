<?php

namespace App\Models;

use Illuminate\Support\Facades\DB;
use Illuminate\Support\Facades\Log;

class Habits {
    static $validationRules = [
        'habit_name' => 'required|string|max255',
        'repeat_type' => 'required|string|in:day,week,day_of_week',
        'repeat_value' => 'required|integer',
        'started_at' => 'required|integer',
        'target_time' => 'nullable|integer',
        'time_of_day' => 'required|string|in:always,am,pm',
        'consecutive_days' => 'required|integer',
        'consecutive_weeks' => 'nullable|integer',
    ];

    static $tableName = 'habits';

    function __construct(){
        $this->pdo = DB::connection()->getPdo();
    }
    
    public function index(){
        $sth = $this->pdo->prepare("select * from ".self::$tableName);
        $sth->execute();
        $result = $sth->fetchAll();
        return $result;
    }

    public function store($params){
        $columns = implode(', ',array_keys($params));
        $values = implode(', ',array_map(function($v){return $v===NULL?"NULL":"'$v'";},array_values($params)));
        $sth = $this->pdo->prepare("insert into ".self::$tableName." (${columns}) values (${values})");
        return $sth->execute();
    }

    public function update($id,$params){
        $setValues = array_map(function($key,$value){
            return $value === NULL ? "$key = NULL" : "$key = '$value'";
        },array_keys($params),array_values($params));
        $setValues = implode(', ',$setValues);
        $sth = $this->pdo->prepare("update ".self::$tableName." set $setValues where id = $id");
        return $sth->execute();
    }

    public function destroy($id){
        $sth = $this->pdo->prepare("delete from ".self::$tableName." where id = ${id}");
        return $sth->execute();
    }
}