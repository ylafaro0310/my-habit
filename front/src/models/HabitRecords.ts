import { List, Record } from 'immutable';
import { Dayjs } from 'dayjs';

import dayjs from '../lib/dayjs-ja';
import { JSObject } from '../types/Common';

export class HabitRecord extends Record<{
  habitId: number;
  completedAt: Dayjs;
  isSkipped: boolean;
}>({
  habitId: 0,
  completedAt: dayjs(),
  isSkipped: false,
}) {
  static fromResponse(response: JSObject): HabitRecord {
    const params = { ...response };
    params['completedAt'] = dayjs(response['completedAt'], 'YYYY-MM-DD HH:mm:ss');
    return new HabitRecord(params);
  }
}

export default class HabitRecords extends Record<{
  items: List<HabitRecord>;
}>({
  items: List(),
}) {
  static fromResponse(response: JSObject): HabitRecords {
    const params = { ...response };
    params.items = List(params.habitRecords.map((item: JSObject) => HabitRecord.fromResponse(item)));
    return new HabitRecords(params);
  }
  getList(): List<HabitRecord> {
    return this.get('items');
  }
  filterById(id: number): HabitRecords {
    return this.set(
      'items',
      this.get('items').filter(habitRecord => habitRecord.habitId === id),
    );
  }
  filterByCompletedAt(date: Dayjs): HabitRecords {
    return this.set(
      'items',
      this.get('items').filter(habitRecord => habitRecord.completedAt.isSame(date, 'd')),
    );
  }
}
