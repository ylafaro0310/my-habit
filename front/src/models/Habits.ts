import { List, Record } from 'immutable';
import dayjs, { Dayjs } from 'dayjs';
import { JSObject } from '../types/Common';
import HabitRecords from './HabitRecords';

export class Habit extends Record<{
    id: number;
    habit_name: string;
    repeat_type: string;
    repeat_value: number;
    started_date: Dayjs;
    target_time: number; // 単位は「分」
    time_of_day: string;
    consecutive_days: number;
    consecutive_weeks: number | null;
}>({
    id: 0,
    habit_name: '',
    repeat_type: 'day_of_week',
    repeat_value: 127,
    started_date: dayjs(),
    target_time: 0,
    time_of_day: '',
    consecutive_days: 0,
    consecutive_weeks: null,
}) {
  static fromResponse(response: JSObject) {
    const params = { ...response };
    return new Habit(params);
  }
  isCompleted(date: Dayjs,habit_records: HabitRecords): boolean {
    return habit_records.filterById(this.id).filterByCompletedAt(date = dayjs('2020-01-08')).getList().size > 0
  }
}

export default class Habits extends Record<{
  items: List<Habit>;
}>({
  items: List(),
}){
  static fromResponse(response: JSObject) {
    const params = { ...response };
    params.items = List(params.habits.map((item: JSObject) => Habit.fromResponse(item)));
    return new Habits(params);
  }
  getList(): List<Habit>{
    return this.get('items');
  }
}