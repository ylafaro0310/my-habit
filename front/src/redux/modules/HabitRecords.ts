import actionCreatorFactory from 'typescript-fsa';
import { reducerWithInitialState } from 'typescript-fsa-reducers';

import HabitRecords from '../../models/HabitRecords';

// Action Creator
const actionCreator = actionCreatorFactory('HabitRecord');
export const HabitRecordsActions = {
  getHabitRecords: actionCreator<object | undefined>('getHabitRecords'),
  setHabitRecords: actionCreator<HabitRecords>('setHabitRecords'),
  addHabitRecord: actionCreator<object>('addHabitRecord'),
  removeHabitRecord: actionCreator<number>('removeHabitRecord'),
};

// Reducers
export const habitRecordsReducer = reducerWithInitialState(new HabitRecords()).case(
  HabitRecordsActions.setHabitRecords,
  (state, payload) => {
    return state.set('items', payload.getList());
  },
);
