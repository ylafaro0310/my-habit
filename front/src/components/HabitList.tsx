import React from 'react';
import styled from 'styled-components';
import dayjs from 'dayjs';

import Habits from '../models/Habits';
import HabitRecords from '../models/HabitRecords';

type HabitListProps = {
    habits: Habits;
    habitRecords: HabitRecords;
};

const HabitList: React.FC<HabitListProps> = ({ habits, habitRecords }: HabitListProps) => {
    return (
        <div className='habits'>
            <form>
                {habits.getList().map((habit, key) => (
                    <ListItem key={key}>
                        <Checkbox
                            id={'habit_' + habit.id}
                            type='checkbox'
                            checked={habit.isCompleted(dayjs('2020-01-08'), habitRecords)}
                        />
                        <label htmlFor={'habit_' + habit.id}>{habit.habitName}</label>
                    </ListItem>
                ))}
            </form>
        </div>
    );
};

const ListItem = styled.div``;

const Checkbox = styled.input``;

export default HabitList;
