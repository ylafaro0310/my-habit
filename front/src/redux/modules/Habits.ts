import actionCreatorFactory from 'typescript-fsa';
import { reducerWithInitialState } from 'typescript-fsa-reducers';

import Habits from '../../models/Habits';

// Action Creator
const actionCreator = actionCreatorFactory('Habit');
export const HabitsActions = {
  getHabits: actionCreator<void>('getHabits'),
  setHabits: actionCreator<Habits>('setHabits'),
  addHabit: actionCreator<object>('addHabit'),
  updateHabit: actionCreator<{ habitId: number; values: object }>(
    'updateHabit',
  ),
  removeHabit: actionCreator<number>('removeHabit'),
  formInitialize: actionCreator<number>('habitFormInitialize'),
};

// Reducers
export const habitsReducer = reducerWithInitialState(new Habits()).case(
  HabitsActions.setHabits,
  (state, payload) => {
    return state.set('items', payload.getList());
  },
);
