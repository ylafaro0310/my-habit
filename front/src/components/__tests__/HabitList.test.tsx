import React from 'react';
import { shallow } from 'enzyme';
import dayjs from 'dayjs';

import Habits from '../../models/Habits';
import HabitRecords from '../../models/HabitRecords';
import { HabitList } from '../HabitList';

const data = {
  habits: [
    { id: 1, habitName: '読書する' },
    { id: 2, habitName: '筋トレ' },
    { id: 3, habitName: '新しいCDを1枚聴く' },
  ],
  habitRecords: [
    { habitId: 2, completedAt: dayjs() },
    { habitId: 1, completedAt: dayjs('2020-01-08') },
  ],
};

const props = {
  habits: Habits.fromResponse(data),
  habitRecords: HabitRecords.fromResponse(data),
  selectedDate: dayjs('2020-01-08').format('YYYY-MM-DD'),
};

describe('Component: HabitList', () => {
  it('Snapshot test', () => {
    const mockFn1 = jest.fn();
    const mockFn2 = jest.fn();
    const wrapper = shallow(<HabitList {...props} addHabitRecord={mockFn1} removeHabitRecord={mockFn2} />);
    expect(wrapper).toMatchSnapshot();
  });
  it('習慣をチェックしたときaddHabitRecordが発行されること', () => {
    const mockFn1 = jest.fn();
    const mockFn2 = jest.fn();
    const wrapper = shallow(<HabitList {...props} addHabitRecord={mockFn1} removeHabitRecord={mockFn2} />);
    wrapper
      .find('input[type="checkbox"]')
      .at(0)
      .simulate('change', { target: { checked: true } });
    expect(mockFn1.mock.calls.length).toBe(1);
  });
  it('習慣のチェックを外したときremoveHabitRecordが発行されること', () => {
    const mockFn1 = jest.fn();
    const mockFn2 = jest.fn();
    const wrapper = shallow(<HabitList {...props} addHabitRecord={mockFn1} removeHabitRecord={mockFn2} />);
    wrapper
      .find('input[type="checkbox"]')
      .at(0)
      .simulate('change', { target: { checked: false } });
    expect(mockFn2.mock.calls.length).toBe(1);
  });
});
