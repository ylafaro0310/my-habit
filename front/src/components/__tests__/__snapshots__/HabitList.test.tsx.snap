// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: HabitList Snapshot test 1`] = `
<StyledCard>
  <form>
    <styled.div
      key="0"
    >
      <input
        checked={true}
        id="habit_1"
        onChange={[Function]}
        type="checkbox"
      />
      <label
        htmlFor="habit_1"
      >
        <div>
          読書する
        </div>
        <div>
          0
          日連続
        </div>
      </label>
      <Link
        to="habits/1"
      >
        <span
          className="fas fa-edit"
        />
      </Link>
    </styled.div>
    <styled.div
      key="1"
    >
      <input
        checked={false}
        id="habit_2"
        onChange={[Function]}
        type="checkbox"
      />
      <label
        htmlFor="habit_2"
      >
        <div>
          筋トレ
        </div>
        <div>
          0
          日連続
        </div>
      </label>
      <Link
        to="habits/2"
      >
        <span
          className="fas fa-edit"
        />
      </Link>
    </styled.div>
    <styled.div
      key="2"
    >
      <input
        checked={false}
        id="habit_3"
        onChange={[Function]}
        type="checkbox"
      />
      <label
        htmlFor="habit_3"
      >
        <div>
          新しいCDを1枚聴く
        </div>
        <div>
          0
          日連続
        </div>
      </label>
      <Link
        to="habits/3"
      >
        <span
          className="fas fa-edit"
        />
      </Link>
    </styled.div>
  </form>
</StyledCard>
`;
